src/NativeFunctionCall.ts

```diff
@@ -0,0 +1,436 @@
+			throw new Error('Unexpected number of parameters');
+		}
+
+		let hasList  = false;
+		for (let p of parameters) {
+			if (p instanceof Void) throw new StoryException('Attempting to perform operation on a void value. Did you forget to "return" a value from a function you called here?');
+			if (p instanceof ListValue)
+				hasList = true;
+		}
+
+		if (parameters.length == 2 && hasList){
+			return this.CallBinaryListOperation(parameters);
+		}
+
+		let coercedParams = this.CoerceValuesToSingleType(parameters);
+		let coercedType = coercedParams[0].valueType;
+
+		if (coercedType == ValueType.Int) {
+			return this.CallType<number>(coercedParams);
+		} else if (coercedType == ValueType.Float) {
+			return this.CallType<number>(coercedParams);
+		} else if (coercedType == ValueType.String) {
+			return this.CallType<string>(coercedParams);
+		} else if (coercedType == ValueType.DivertTarget) {
+			return this.CallType<Path>(coercedParams);
+		} else if (coercedType == ValueType.List) {
+			return this.CallType<InkList>(coercedParams);
+		}
+
+		return null;
+	}
+
+	public CallType<T>(parametersOfSingleType: Array<Value<T>>){
+		let param1 = asOrThrows(parametersOfSingleType[0], Value);
+		let valType = param1.valueType;
+
+		let val1 = param1 as Value<T>;
+
+		let paramCount = parametersOfSingleType.length;
+
+		if (paramCount == 2 || paramCount == 1) {
+			if (this._operationFuncs === null) return throwNullException('NativeFunctionCall._operationFuncs');
+			let opForTypeObj = this._operationFuncs.get(valType);
+			if (!opForTypeObj) {
+				throw new StoryException('Cannot perform operation '+this.name+' on '+valType);
+			}
+
+			if (paramCount == 2) {
+				let param2 = asOrThrows(parametersOfSingleType[1], Value);
+
+				let val2 = param2 as Value<T>;
+
+				let opForType = opForTypeObj as BinaryOp<T>;
+
+				if (val1.value === null || val2.value === null) return throwNullException('NativeFunctionCall.Call BinaryOp values');
+				let resultVal = opForType(val1.value, val2.value);
+
+				return Value.Create(resultVal);
+			}
+
+			else {
+
+				let opForType = opForTypeObj as UnaryOp<T>;
+
+				if (val1.value === null) return throwNullException('NativeFunctionCall.Call UnaryOp value');
+				let resultVal = opForType(val1.value);
+
+				return Value.Create(resultVal);
+			}
+		}
+
+		else {
+			throw new Error('Unexpected number of parameters to NativeFunctionCall: ' + parametersOfSingleType.length);
+		}
+	}
+
+	public CallBinaryListOperation(parameters: InkObject[]){
+		if ((this.name == '+' || this.name == '-') && parameters[0] instanceof ListValue && parameters[1] instanceof IntValue)
+			return this.CallListIncrementOperation(parameters);
+
+		let v1 = asOrNull(parameters[0], Value);
+		let v2 = asOrNull(parameters[1], Value);
+
+		if (v1 === null) return throwNullException('NativeFunctionCall.CallBinaryListOperation v1');
+		if (v2 === null) return throwNullException('NativeFunctionCall.CallBinaryListOperation v2');
```

Comment:
We should probably use `asOrThrows` directly here, instead of `asOrNull` followed by a manual null check. It does make the code slightly different, but not by much. Nat did something similar in VariablesStates and I found the change to be more elegant in the end.

As long as we explain the rationals in a CONTRIBUTING.md, we'll be fine, what do you think?

Reply:
Sounds good!